tool/src/c2/ty.rs

```diff
@@ -256,31 +215,34 @@ impl<'ccx, 'tcx: 'ccx, 'header> TyGenContext<'ccx, 'tcx, 'header> {
         }
     }
 
-    pub fn gen_result(&mut self, name: &str, ty: ResultType) {
-        let ok_line = if let Some(ok) = ty.0 {
-            let ok_name = self.gen_ty_name(ok, true);
-            format!("\t\t{ok_name} ok;\n")
+    pub fn gen_result_ty(
+        &mut self,
+        fn_name: &str,
+        ok_ty: Option<&hir::OutType>,
+        err_ty: Option<&hir::OutType>,
+    ) -> String {
+        let ok_line = if let Some(ok) = ok_ty {
+            let ok_name = self.gen_ty_name(ok, false);
+            format!("{ok_name} ok;")
         } else {
             "".into()
         };
-        let err_line = if let Some(err) = ty.1 {
-            let err_name = self.gen_ty_name(err, true);
-            format!("\t\t{err_name} err;\n")
+        let err_line = if let Some(err) = err_ty {
+            let err_name = self.gen_ty_name(err, false);
+            format!("{err_name} err;")
         } else {
             "".into()
         };
 
-        let union_def = if ty.0.is_some() || ty.1.is_some() {
-            format!("\tunion {{\n{ok_line}{err_line}\t}};\n")
+        let union_def = if ok_ty.is_some() || err_ty.is_some() {
+            format!("union {{{ok_line} {err_line}}};")
         } else {
             "".into()
         };
 
-        writeln!(
-            self.decl_header,
-            "typedef struct {name} {{\n{union_def}\tbool is_ok;\n}} {name};"
-        )
-        .unwrap();
+        // We can't use an anonymous struct here: C++ doesn't like producing those in return types
+        // Instead we name it something unique per-function. This is a bit ugly but works just fine.
+        format!("struct {fn_name}_result {{{union_def} bool is_ok;}};\nstruct {fn_name}_result")
```

Comment:
This currently generates a tagged struct (`struct fnname_result {..}`). We could instead generate a type: `typedef struct fnname_result {} fnname_result;`

I picked a tag because this reduces the amount of namespace pollution, polluting the set of tags but not the set of types (and reducing potentials for confusing the diagnostics).


But open to doing it the other way

Reply:
I think I lean toward `typedef` because that's more standard, but I'm fine with exploring a tagged struct due to less namespace polution.

Reply:
We could also not define a type at all and put `struct { union {...}; bool is_ok }` directly in the return type. Also no namespace pollution.

Reply:
No, we cannot, C++ doesn't let you do that (C does)

That was the origin plan for this but I had to give it up.